### ISSUE TITLE
Are traceparent fields optional or fixed width?

### ISSUE THREAD

**Issue by codefromthecrypt** [w3c/trace-context#115]
> When `trace-options` is missing the default value for this bit is `0`.

This part of the http doc hints that the field could be left out. Are any of the fields something we can leave out? Or is this a mistake?

**Comment by nicmunroe** [comment]
Hmm. Having fields optional leaves ambiguity: did the setter intend for it to be missing or did they mess up? I'm not sure if there are any tracing systems that have no concept of trace ID or span ID (even if they call it something different), but if there are then the recommendation could be to create random values, or all 0's, or something - point being we could call it out in the spec what to do so it's explicit. The trace options field already has `0` defined as an "I don't care and am deferring to downstream", so that's covered.

Without hearing reasons for wanting fields to be optional I'd lean towards making them all required. The only reason I can think of is passing a few less bytes on the request, but I'm not sure if that argument would sway me.

TL;DR - I'm open to arguments for why fields should be optional, but at the moment I'm leaning towards making them all required, mainly to eliminate ambiguity.

**Comment by nicmunroe** [comment]
Requiring all fields should also make the spec simpler and smaller - no need to define and explain multiple ways of doing the same things, parsing rules, etc.

**Comment by SergeyKanzhelev** [comment]
only trace options are optional: https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md#field-value I think the rationale for that was that some vendors do not want to specify any trace options. Basically saving on 3 bytes which has little meaning for their systems. I don't remember which vendors were that.

One proposal we declined was to allow `trace-id`-only `traceparent` headers. The intended use was to save size for small clients (like mobile devices) initiating the call. Rationale for declining it was to avoid abuse and confusion. Suggestion on saving size is to use binary format that we want to discuss.

I'd be ok to make trace options required as well.


BTW, @nicmunroe - all `0` considered an invalid span and trace ID.

**Comment by nicmunroe** [comment]
Thanks for the explanation. My perspective is pretty narrow, but without more arguments from people or vendors who actively want to save those 3 bytes, I'd vote to make trace options required.

**Comment by SergeyKanzhelev** [comment]
three people have consensus. I created PR to get more feedback. Maybe it is not an issue any longer.

**Comment by SergeyKanzhelev** [comment]
closing as we all agree on fixed length
